simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,427 @@
+                    threshold /= reg.regularization_mesh.base_length
+
+                obj.irls_threshold = threshold
+
+            reg.norms = norms
+
+            if self.verbose:
+                print("irls_threshold " + str(reg.objfcts[0].irls_threshold))
+
+        # Save l2-model
+        self.invProb.l2model = self.invProb.model.copy()
+
+    def validate(self, directiveList=None):
+        directive_list = directiveList.dList
+        self_ind = directive_list.index(self)
+        lin_precond_ind = [isinstance(d, UpdatePreconditioner) for d in directive_list]
+
+        if any(lin_precond_ind):
+            assert lin_precond_ind.index(True) > self_ind, (
+                "The directive 'UpdatePreconditioner' must be after Update_IRLS "
+                "in the directiveList"
+            )
+        else:
+            warnings.warn(
+                "Without a Linear preconditioner, convergence may be slow. "
+                "Consider adding `Directives.UpdatePreconditioner` to your "
+                "directives list",
+                stacklevel=2,
+            )
+
+        beta_schedule = [
+            d for d in directive_list if isinstance(d, BetaSchedule) and d != self
+        ]
+
+        assert len(beta_schedule) == 0, (
+            "Beta scheduling is handled by the `UpdateIRLS` directive."
+            "Remove the redundant `BetaSchedule` from your list of directives.",
+        )
+
+        spherical_scale = [isinstance(d, SphericalDomain) for d in directive_list]
+        if any(spherical_scale):
+            assert spherical_scale.index(True) < self_ind, (
+                "The directive 'SphericalDomain' must be before UpdateIRLS "
+                "in the directiveList"
+            )
+
+        return True
+
+    def stopping_criteria(self):
+        """
+        Check for stopping criteria of max_irls_iteration or minimum change.
+        """
+        phim_new = 0
+        for reg in self.reg.objfcts:
+            if isinstance(reg, (Sparse, BaseSparse)):
+                reg.model = self.invProb.model
+                phim_new += reg(reg.model)
+
+        # Check for maximum number of IRLS cycles
+        if self.metrics.irls_iteration_count == self.max_irls_iterations:
+            if self.verbose:
+                print(
+                    "Reach maximum number of IRLS cycles:"
+                    + " {0:d}".format(self.max_irls_iterations)
+                )
+            return True
+
+        # Check if the function has changed enough
+        f_change = np.abs(self.metrics.f_old - phim_new) / (self.metrics.f_old + 1e-12)
+
+        if (
+            f_change < self.f_min_change
+            and self.metrics.irls_iteration_count > 1
+            and np.abs(
+                1.0
+                - self.invProb.phi_d / self.misfit_from_chi_factor(self.chifact_target)
+            )
+            < self.misfit_tolerance
+        ):
+            print("Minimum decrease in regularization." + "End of IRLS")
```

Comment:
Should we add a `if self.verbose` here?
```suggestion
            if self.verbose:
                print("Minimum decrease in regularization." + "End of IRLS")
```